packages/flutter_tools/lib/src/web/compiler_config.dart

```diff
@@ -13,22 +13,29 @@ enum CompileTarget {
 
 sealed class WebCompilerConfig {
   const WebCompilerConfig({required this.renderer,
-                           required this.optimizationLevel,
+                           this.optimizationLevel,
                            required this.sourceMaps});
-
-  /// The default optimization level for dart2js/dart2wasm.
-  static const int kDefaultOptimizationLevel = 4;
-
   /// Build environment flag for [optimizationLevel].
   static const String kOptimizationLevel = 'OptimizationLevel';
 
   /// Build environment flag for [sourceMaps].
   static const String kSourceMapsEnabled = 'SourceMaps';
 
-  /// The compiler optimization level.
+  /// Calculates the optimization level for dart2js/dart2wasm for the given
+  /// build mode.
+  int optimizationLevelForBuildMode(BuildMode mode) =>
+    optimizationLevel ?? switch (mode) {
+      BuildMode.debug => 0,
+      BuildMode.profile || BuildMode.release => 4,
+      BuildMode.jitRelease => throw ArgumentError('Invalid build mode for web'),
+    };
+
+  /// The compiler optimization level specified by the user.
   ///
   /// Valid values are O0 (lowest, debug default) to O4 (highest, release default).
-  final int optimizationLevel;
+  /// If the value is null, the user hasn't specified an optimization level and we
+  /// will use the default appropriate for teh build mode.
```

Comment:
nit 1: avoid "[we](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-the-passive-voice-recommend-do-not-require-never-say-things-are-simple)" in the docs. Maybe something like "...and the appropriate default should be used."

nit 2: s/teh/the/